### PR TITLE
staging: add budgets helpers and more validations

### PR DESCRIPTION
**Pull request by njtran**
Fixes #N/A 

**Description**
This adds helpers that are going to be used when considering disruption budgets for disruption. Adds more CEL validations on the values of durations.

**How was this change tested?**
make presubmit

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/v1beta1/nodepool.go [hunk 1]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
...
+	}
+	// return the values, defaulting to -1 if the value is MaxInt64
+	return intstr.FromInt(lo.Ternary(minIntVal == math.MaxInt64, -1, minIntVal)), intstr.FromString(fmt.Sprintf("%d%%", lo.Ternary(minPercentVal == math.MaxInt64, -1, minPercentVal))), nil
+}
+
+// GetAllowedDisruptions returns an intstr.IntOrString that can be used a comparison
+// for calculating if a disruption action is allowed. It returns an error if the
+// crontab is invalid. This returns -1 if the value is unbounded.
+func (b *Budget) GetAllowedDisruptions(c clock.Clock) (intstr.IntOrString, error) {
+	active, err := b.IsActive(c)
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:256]: When does this error? Is there ever a valid case where this errors out?

File: pkg/apis/v1beta1/nodepool.go [hunk 2]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
 		return ptr.Int32Value(pl.Items[a].Spec.Weight) > ptr.Int32Value(pl.Items[b].Spec.Weight)
 	})
 }
+
+// GetAllowedDisruptions returns the minimum allowed disruptions across all disruption budgets for a given node pool.
+// This returns two values as the resolved value for a percent depends on the number of current node claims.
+func (in *NodePool) GetAllowedDisruptions(ctx context.Context, c clock.Clock) (intstr.IntOrString, intstr.IntOrString, error) {
+	vals := make([]intstr.IntOrString, len(in.Spec.Disruption.Budgets))
+	errs := make([]error, len(in.Spec.Disruption.Budgets))
+	workqueue.ParallelizeUntil(ctx, len(in.Spec.Disruption.Budgets), len(in.Spec.Disruption.Budgets), func(i int) {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:219]: Any particular reason that we want to parallelize this? Do we expect this to be complicated or take a long time?

File: pkg/apis/v1beta1/nodepool.go [hunk 3]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
 		return ptr.Int32Value(pl.Items[a].Spec.Weight) > ptr.Int32Value(pl.Items[b].Spec.Weight)
 	})
 }
+
+// GetAllowedDisruptions returns the minimum allowed disruptions across all disruption budgets for a given node pool.
+// This returns two values as the resolved value for a percent depends on the number of current node claims.
+func (in *NodePool) GetAllowedDisruptions(ctx context.Context, c clock.Clock) (intstr.IntOrString, intstr.IntOrString, error) {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:216]: I wasn't quite getting this from the description, but my assumption is that one of these values is going to be a percentage value and the other is going to be an integer value. Is there a reason that they are both `IntOrString` since one should always be a string and the other should always be an integer?

File: pkg/apis/v1beta1/nodepool.go [hunk 4]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
...
+		vals[i] = val
+	})
+	if err := multierr.Combine(errs...); err != nil {
+		return intstr.IntOrString{}, intstr.IntOrString{}, err
+	}
+	minIntVal, minPercentVal := math.MaxInt64, math.MaxInt64
+	for i := range vals {
+		val := vals[i]
+		// The crontab wasn't active
+		if val.IntVal == -1 {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:233]: Does -1 mean not active or does -1 mean infinite? If I look below, I thought that you had mentioned that -1 was infinite. I was expecting an inactive budget to have an integer value of 0

File: pkg/apis/v1beta1/nodepool.go [hunk 5]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
...
+			continue
+		}
+		// This returns the percent value if it's a string, and the raw value if it's an int.
+		temp, err := intstr.GetScaledValueFromIntOrPercent(lo.ToPtr(val), 100, false)
+		if err != nil {
+			// Should almost never happen since this is validated when the nodepool is applied
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:239]: Can we guarantee that this should never happen? I'd like to avoid having errors that "pretty much can't ever happen" since I question why we should be handling the error in the code at all

File: pkg/apis/v1beta1/nodepool.go (hunk 3 above)
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:216]: Rather than handling everything as `IntOrString` does it make sense to just divide up the budgets into percentage-based ones and integer-based ones and take the minimum that way rather than handling both cases simultaneously and trying to split it later?

File: pkg/apis/v1beta1/nodepool.go [hunk 6]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
...
+// If the last crontab hit is exactly the duration in the past, this means the
+// schedule is active, as any more crontab hits in between would only extend this
+// window. This ensures that any previous crontab hits for a schedule are considered.
+func (b *Budget) IsActive(c clock.Clock) (bool, error) {
+	if b.Crontab == nil && b.Duration == nil {
+		return true, nil
+	}
+	schedule, err := cron.ParseStandard(lo.FromPtr(b.Crontab))
+	if err != nil {
+		return false, fmt.Errorf("parsing crontab, %w", err)
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:275]: Same question here. Should this ever fail?

File: pkg/apis/v1beta1/nodepool.go [hunk 7]
```diff
@@ -202,3 +210,72 @@ func (pl *NodePoolList) OrderByWeight() {
...
+	}
+	// Walk back in time for the duration associated with the crontab
+	checkPoint := c.Now().Add(-lo.FromPtr(b.Duration).Duration)
+	nextHit := schedule.Next(checkPoint)
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:279]: Is there a clearer way to do this logic? I get how it's working but the way it's written makes it a little hard to follow. Perhaps an example in the comment could help clear it up but I'm also curious if there's a concept of "between" where we can get the last crontab and see if we are currently between that time and the duration after it

File: pkg/apis/v1beta1/nodepool_test.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
+/*
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool_test.go:1]: optional: Should this file be `nodepool_budgets_test`?

File: pkg/apis/v1beta1/nodepool_validation_cel_test.go [hunk 9]
```diff
@@ -105,15 +105,49 @@ var _ = Describe("CEL/Validation", func() {
 			nodePool.Spec.Disruption.Budgets = []Budget{{
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool_validation_cel_test.go:105]: Should we be making the same test changes to the `nodepool_validation_webhook_test`?

File: pkg/apis/v1beta1/nodepool.go (hunk 2 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:219]: Was just doing this to make it quicker and do the multierr handling like we do elsewhere. Don't mind moving it to a for loop with err.Append() calls if that's what you prefer.

File: pkg/apis/v1beta1/nodepool.go (hunk 3 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:216]: The IntOrString has a Type value that allows the GetScaledValue to infer whether it needs to do a percentage or a raw value. I could probably put both of the values in the return value here, but we'd need to in-place modify the type to compare the two resolved values together, where I think having two values may just more deliberate here.

File: pkg/apis/v1beta1/nodepool.go (hunk 4 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:233]: not active and infinite is the same. If a schedule isn't active, there's no budget that should be applied. An inactive budget defaulting to 0 means that when the budget isn't active, no disruption can happen, which is the opposite of what we want here.

File: pkg/apis/v1beta1/nodepool.go (hunk 5 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:239]: If the validation mechanism isn't able to validate/default an object for some reason, a user could get themselves into config error. What are your thoughts on this?

File: pkg/apis/v1beta1/nodepool.go (hunk 3 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:216]: I could see an argument for this. This is essentially what I do for the comparison in this function, but if we don't return an intstr.IntOrString() here, we have to create a new one in the caller function to then create the intstr.IntOrString() as an input to GetScaledValue().

File: pkg/apis/v1beta1/nodepool.go (hunk 7 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:279]: If you have a suggestion on how to make it clearer in code, I'm open to it :D 

Let me know if this example helps:
```
Crontab = @hourly == 0 * * * * and time.Now() = 10:15 and duration = 30m, we'll walk back to 9:45, get the next crontab hit from there, which should be 10, which is before 10:15, meaning the cron is active. 
Doing this with the same but time.Now() = 10:45, we walk back in time to 10:15 and then get the next crontab hit, which is 11:00. Since this is after 10:45, we say that the cron is not active.
```

File: pkg/apis/v1beta1/nodepool_validation_cel_test.go (hunk 9 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool_validation_cel_test.go:105]: Will do in a follow up when I add those tests.

Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodepool_validation_cel_test.go:105]: I see. So we're not adding them right now because we don't yet have the webhook validations. Makes sense.

File: pkg/apis/v1beta1/nodepool.go (hunk 7 above)
Comment (garvinp-stripe) [reply pkg/apis/v1beta1/nodepool.go:279]: I think (maybe thats just me) the naming can be a bit confusing. Rather than calling it `nextHit` maybe it should be something like `closestLastHit`?

Comment (garvinp-stripe) [reply pkg/apis/v1beta1/nodepool.go:279]: I think its difficult given the cron parser only gives you `Next` based on a time and finding the last crontab is quite hard to do. I think this method make sense even though is slightly complicated to understand

File: pkg/apis/v1beta1/nodepool.go (hunk 1 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:256]: It errors out when the the cron parser catches something that validation didn't. Since regex is notorious for having edge cases, I'd like to keep this here regardless of how fleshed out our validation is to ensure that we're always getting a schedule that's valid.

File: pkg/apis/v1beta1/nodepool.go (hunk 6 above)
Comment (njtran) [reply pkg/apis/v1beta1/nodepool.go:275]: If there's an edge case with our regex that the cron parser catches but we don't, then yes.

**Comment by njtran** [comment]
closed and reopened to re-trigger presubmits that weren't showing up

File: pkg/apis/v1beta1/nodepool.go [hunk 10]
```diff
@@ -98,20 +104,25 @@ type Budget struct {
 	// MaxUnavailable dictates how many NodeClaims owned by this NodePool
 	// can be terminating at once. It must be set.
 	// This only considers NodeClaims with the karpenter.sh/disruption taint.
-	// +kubebuilder:validation:XIntOrString
+	// We can't use an intstr.IntOrString since kubebuilder doesn't support pattern
+	// checking for int values for IntOrString values.
+	// Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+	// +kubebuilder:validation:Type="string"
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:110]: Do you need this validation type now that you are explicitly defining this value as a string?

File: pkg/apis/v1beta1/nodepool.go [hunk 11]
```diff
@@ -199,3 +210,79 @@ func (pl *NodePoolList) OrderByWeight() {
...
+			minPercentVal = lo.Ternary(temp < minPercentVal, temp, minPercentVal)
+		}
+	}
+	// return the values, defaulting to -1 if the value is MaxInt64
+	return intstr.FromInt(lo.Ternary(minIntVal == math.MaxInt64, -1, minIntVal)), intstr.FromString(fmt.Sprintf("%d%%", lo.Ternary(minPercentVal == math.MaxInt64, -1, minPercentVal))), nil
+}
+
+// GetAllowedDisruptions returns an intstr.IntOrString that can be used a comparison
+// for calculating if a disruption action is allowed. It returns an error if the
+// crontab is invalid. This returns -1 if the value is unbounded.
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool.go:254]: I wonder if it makes more sense to use MAX_INT here so that what's stored in the value is more indicative of what's actually happening here. Otherwise, we are having to reason about what `-1` means here. Thoughts?

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀
